Tools/TestWebKitAPI/Tests/WebKitGtk/TestWebKitAccessibility.cpp

```diff
@@ -641,6 +641,40 @@ static void testAccessibleState(AccessibilityTest* test, gconstpointer)
     g_assert_true(atspi_state_set_contains(stateSet.get(), ATSPI_STATE_SHOWING));
 }
 
+static void testAccessibleStateChangedFocus(AccessibilityTest* test, gconstpointer)
+{
+    test->showInWindow(800, 600);
+    test->loadHtml(
+        "<html>"
+        "  <body>"
+        "    <div id='container'></div>"
+        "  </body>"
+        "</html>",
+        nullptr);
+    test->waitUntilLoadFinished();
+
+    auto testApp = test->findTestApplication();
+    g_assert_true(ATSPI_IS_ACCESSIBLE(testApp.get()));
+
+    auto documentWeb = test->findDocumentWeb(testApp.get());
+    g_assert_true(ATSPI_IS_ACCESSIBLE(documentWeb.get()));
+
+    test->startEventMonitor(nullptr, { "object:state-changed" });
+    test->runJavaScriptAndWaitUntilFinished(
+        "container=document.getElementById('container');"
+        "var i = document.createElement('div');"
+        "i.setAttribute('tabindex', '-1');"
+        "container.appendChild(i);"
+        "i.innerHTML = 'TEST';"
+        "i.focus();",
+        nullptr);
+    auto events = test->stopEventMonitor(1);
+    g_assert_cmpuint(events.size(), ==, 1);
+    g_assert_cmpstr(events[0]->type, ==, "object:state-changed:focused");
+    auto* div = events[0]->source;
+    g_assert_true(ATSPI_IS_ACCESSIBLE(div));
+}
```

Comment:
It feels unfortunate that the bug you're fixing is in WebCore (vs. the platform layer), but the test being added is GTK only. If this is a bug in WebCore, then the bad / incomplete state situation you're fixing should be observable from a layout test, using the platform-agnostic AccessibilityUIElement APIs.

Also, what is the incomplete / bad state that manifests when this bug happens? Is it that the object is erroneously ignored (a.k.a `!ATSPI_IS_ACCESSIBLE(div)`)? Or something else?

Reply:
You are right, it should be possible to create a Layout test for that. I will try to do that.